triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala

```diff
@@ -56,27 +56,28 @@ object TriggerServiceFixture {
     val host = InetAddress.getLoopbackAddress
     val isWindows: Boolean = sys.props("os.name").toLowerCase.contains("windows")
 
-    // Launch a toxiproxy instance. Wait on it to be ready to accept
-    // connections.
-    val toxiProxyExe =
-      if (!isWindows)
-        BazelRunfiles.rlocation("external/toxiproxy_dev_env/bin/toxiproxy-cmd")
-      else
-        BazelRunfiles.rlocation("external/toxiproxy_dev_env/toxiproxy-server-windows-amd64.exe")
-    val toxiProxyPort = LockedFreePort.find()
-    val toxiProxyProc =
-      Process(Seq(toxiProxyExe, "--port", toxiProxyPort.port.value.toString)).run()
-    RetryStrategy.constant(attempts = 3, waitTime = 2.seconds)((_, _) =>
-      Future(toxiProxyPort.testAndUnlock(host)))
-    val toxiProxyClient = new ToxiproxyClient(host.getHostName, toxiProxyPort.port.value)
+    // Launch a Toxiproxy server. Wait on it to be ready to accept connections and
+    // then create a client.
+    val toxiproxyExe =
+      if (!isWindows) BazelRunfiles.rlocation("external/toxiproxy_dev_env/bin/toxiproxy-cmd")
+      else BazelRunfiles.rlocation("external/toxiproxy_dev_env/toxiproxy-server-windows-amd64.exe")
+    val toxiproxyF: Future[(Process, ToxiproxyClient)] = for {
+      toxiproxyPort <- Future(LockedFreePort.find())
```

Comment:
It looks like you are now starting toxiproxy in parallel with Sandbox. I believe that is racy, consider the following scenario:

1. `LockedFreePort.find()` will bind to port 0 to get a free port X and immediately stop listening on that port and give you back the port X it got allocated by the OS.
2. Now port X is free, Sandbox starts up bindings to port 0. This doesn’t go through the locking mechanism afaik as it is only used in the test library. The OS is free to allocate Sandbox to port X since nothing is listening on this.
3. Now toxiproxy is started on port X which is already occupied by Sandbox.

At least on Linux, this is very hard to trigger since the Kernel will not immediately reuse the port so Sandbox will be allocated to a different port. Not quite sure about other platforms but I don’t recall seeing issues with this there either. So we probably don’t have to worry about this for now but it’s worth keeping in mind. You could fix it fairly easily by sequencing `toxiproxyF` before starting Sandbox or only create the future after you got `ledgerPort`.

Reply:
Ahh interesting. I'll keep an eye on it and possibly refactor later.

Reply:
Something like this maybe?

```
- should give an 'unauthorized' response for a stop request without an authorization header *** FAILED *** (123 milliseconds)
  java.lang.NullPointerException:
  at com.daml.ports.PortLock$Locked.unlock(PortLock.scala:55)
  at com.daml.ports.PortLock$.lock(PortLock.scala:41)
  at com.daml.ports.LockedFreePort$.find(LockedFreePort.scala:15)
  at com.daml.lf.engine.trigger.TriggerServiceFixture$.$anonfun$withTriggerService$1(TriggerServiceFixture.scala:65)
```

Reply:
That could be it! I'll submit the PR sequencing `toxiproxyF` before the sandbox as it's an easy change anyway.

Reply:
#6939 for the one, and 69d5d87c904add3ee19566f7ce61d00d3c34db3c for the other.